File: java/HeuristicGcode.java
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;
import java.util.Stack;
import java.util.logging.Logger;

public class HeuristicGcode {
    private static final Logger logger = Logger.getLogger(HeuristicGcode.class.getName());

    public static void generate(BufferedImage image, String gcodePath, double resizeFactor) {
        try {
            int width = image.getWidth();
            int height = image.getHeight();
            
            Set<String> pixelSet = new HashSet<>();
            for (int y = 0; y < height; y++) {
                for (int x = 0; x < width; x++) {
                    int pixel = image.getRGB(x, y) & 0xFF;
                    if (pixel > 127) {
                        pixelSet.add(x + "," + y);
                    }
                }
            }
            
            Set<String> visited = new HashSet<>();
            Stack<String> stack = new Stack<>();
            String startPixel = pixelSet.iterator().next();
            stack.push(startPixel);
            
            StringBuilder gcode = new StringBuilder();
            gcode.append("G21 ; Set units to mm\n");
            gcode.append("G90 ; Absolute positioning\n");
            
            try (FileWriter writer = new FileWriter(new File(gcodePath))) {
                while (!stack.isEmpty()) {
                    String currentPixel = stack.pop();
                    if (visited.contains(currentPixel)) continue;
                    visited.add(currentPixel);
                    
                    String[] parts = currentPixel.split(",");
                    int x = Integer.parseInt(parts[0]);
                    int y = Integer.parseInt(parts[1]);
                    gcode.append(String.format("G1 X%.2f Y%.2f F1000\n", x * resizeFactor, y * resizeFactor));
                    
                    for (int dx = -1; dx <= 1; dx++) {
                        for (int dy = -1; dy <= 1; dy++) {
                            if (dx == 0 && dy == 0) continue;
                            String neighbor = (x + dx) + "," + (y + dy);
                            if (pixelSet.contains(neighbor) && !visited.contains(neighbor)) {
                                stack.push(neighbor);
                            }
                        }
                    }
                }
                
                writer.write(gcode.toString());
            }
            
            logger.info("G-code successfully generated: " + gcodePath);
        } catch (IOException e) {
            logger.severe("Error generating G-code: " + e.getMessage());
        }
    }
}

File: python/heuristicGcode.py
import cv2
import numpy as np
import logging

# Set up logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

def Heuristic_gcode(image, gcode_path, resize_factor=1.0):
    logging.info("Starting Heuristic_gcode function")
    
    # Flip the image upside down
    image = np.flipud(image)
    logging.info("Image flipped upside down")

    if len(image.shape) == 3:
        image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        logging.info("Converted image to grayscale")

    # Get coordinates of non-zero pixels
    # y_coords, x_coords = np.nonzero(image)
    y_coords, x_coords = np.nonzero(image > 127)
    points = np.column_stack((y_coords, x_coords))
    logging.info(f"Found {len(points)} non-zero pixels")

    # Create a graph representation of the non-zero pixels
    pixel_set = set(map(tuple, points))
    visited = set()
    path = []

    def dfs(point):
        stack = [point]
        while stack:
            p = stack.pop()
            path.append(p)
            if p in visited:
                continue
            visited.add(p)
            # Check all 8 possible directions (adjacent pixels)
            for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]:
                neighbor = (p[0] + dx, p[1] + dy)
                if neighbor in pixel_set and neighbor not in visited:
                    stack.append(neighbor)
        logging.info(f"DFS completed with path length: {len(path)}")

    # Start DFS from the first nonzero pixel
    dfs(tuple(points[0]))

    # Write G-code
    with open(gcode_path, 'w') as gcode_file:
        gcode_file.write("G21 ; Set units to mm\n")
        gcode_file.write("G90 ; Absolute positioning\n")
        # Move to start position
        gcode_file.write(f"G0 X{path[0][1] * resize_factor:.2f} Y{path[0][0] * resize_factor:.2f}\n")
        for p in path:
            gcode_file.write(f"G1 X{p[1] * resize_factor:.2f} Y{p[0] * resize_factor:.2f} F1800\n")
    logging.info(f"G-code written to {gcode_path}")

    # Plot the image
    # plt.imshow(image, cmap='gray')

    # # Draw the path
    # path = np.array(path)
    # plt.plot(path[:, 1], path[:, 0], 'r-', linewidth=0.8)

    # plt.show()


# Example usage
if __name__ == "__main__":
    image = cv2.imread('resources/EiffelTower.jpg', cv2.IMREAD_GRAYSCALE)
    image = cv2.bitwise_not(image)
    Heuristic_gcode(image, 'output/heuristic_EiffelTower.gcode')
